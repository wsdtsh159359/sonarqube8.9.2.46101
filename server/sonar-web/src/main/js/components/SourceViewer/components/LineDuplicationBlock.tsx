/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import { DropdownOverlay } from 'sonar-ui-common/components/controls/Dropdown';
import Toggler from 'sonar-ui-common/components/controls/Toggler';
import Tooltip from 'sonar-ui-common/components/controls/Tooltip';
import { PopupPlacement } from 'sonar-ui-common/components/ui/popups';
import { translate } from 'sonar-ui-common/helpers/l10n';

export interface LineDuplicationBlockProps {
  blocksLoaded: boolean;
  duplicated: boolean;
  index: number;
  line: T.SourceLine;
  onClick?: (line: T.SourceLine) => void;
  renderDuplicationPopup: (index: number, line: number) => React.ReactNode;
}

export function LineDuplicationBlock(props: LineDuplicationBlockProps) {
  const { blocksLoaded, duplicated, index, line } = props;
  const [dropdownOpen, setDropdownOpen] = React.useState(false);

  const className = classNames('source-meta', 'source-line-duplications', {
    'source-line-duplicated': duplicated
  });

  return duplicated ? (
    <td className={className} data-index={index} data-line-number={line.line}>
      <Tooltip
        overlay={dropdownOpen ? undefined : translate('source_viewer.tooltip.duplicated_block')}
        placement="right">
        <div>
          <Toggler
            onRequestClose={() => setDropdownOpen(false)}
            open={dropdownOpen}
            overlay={
              <DropdownOverlay placement={PopupPlacement.RightTop}>
                {props.renderDuplicationPopup(index, line.line)}
              </DropdownOverlay>
            }>
            <div
              aria-label={translate('source_viewer.tooltip.duplicated_block')}
              className="source-line-bar"
              onClick={() => {
                setDropdownOpen(true);
                if (!blocksLoaded && line.duplicated && props.onClick) {
                  props.onClick(line);
                }
              }}
              role="button"
              tabIndex={0}
            />
          </Toggler>
        </div>
      </Tooltip>
    </td>
  ) : (
    <td className={className} data-index={index} data-line-number={line.line}>
      <div className="source-line-bar" />
    </td>
  );
}

export default React.memo(LineDuplicationBlock);
