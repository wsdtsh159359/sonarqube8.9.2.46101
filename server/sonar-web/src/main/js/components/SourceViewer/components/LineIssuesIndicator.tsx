/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import { uniq } from 'lodash';
import * as React from 'react';
import Tooltip from 'sonar-ui-common/components/controls/Tooltip';
import IssueIcon from 'sonar-ui-common/components/icons/IssueIcon';
import { translate, translateWithParameters } from 'sonar-ui-common/helpers/l10n';
import { sortByType } from '../../../helpers/issues';

export interface LineIssuesIndicatorProps {
  issues: T.Issue[];
  issuesOpen?: boolean;
  line: T.SourceLine;
  onClick: () => void;
}

export function LineIssuesIndicator(props: LineIssuesIndicatorProps) {
  const { issues, issuesOpen, line } = props;
  const hasIssues = issues.length > 0;
  const className = classNames('source-meta', 'source-line-issues', {
    'source-line-with-issues': hasIssues
  });

  if (!hasIssues) {
    return <td className={className} data-line-number={line.line} />;
  }

  const mostImportantIssue = sortByType(issues)[0];
  const issueTypes = uniq(issues.map(i => i.type));

  let tooltipContent;
  if (issueTypes.length > 1) {
    tooltipContent = translate('source_viewer.issues_on_line.multiple_issues');
  } else if (issues.length === 1) {
    tooltipContent = translateWithParameters(
      'source_viewer.issues_on_line.issue_of_type_X',
      translate('issue.type', mostImportantIssue.type)
    );
  } else {
    tooltipContent = translateWithParameters(
      'source_viewer.issues_on_line.X_issues_of_type_Y',
      issues.length,
      translate('issue.type', mostImportantIssue.type, 'plural')
    );
  }

  return (
    <td className={className} data-line-number={line.line}>
      <span
        aria-label={translate('source_viewer.issues_on_line', issuesOpen ? 'hide' : 'show')}
        onClick={(e: React.MouseEvent<HTMLElement>) => {
          e.preventDefault();
          e.currentTarget.blur();
          props.onClick();
        }}
        role="button"
        tabIndex={0}>
        <Tooltip overlay={tooltipContent}>
          <IssueIcon type={mostImportantIssue.type} />
        </Tooltip>
        {issues.length > 1 && <span className="source-line-issues-counter">{issues.length}</span>}
      </span>
    </td>
  );
}

export default React.memo(LineIssuesIndicator);
