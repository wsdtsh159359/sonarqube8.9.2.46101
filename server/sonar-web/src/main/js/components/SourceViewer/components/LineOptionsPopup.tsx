/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { ActionsDropdownItem } from 'sonar-ui-common/components/controls/ActionsDropdown';
import { DropdownOverlay } from 'sonar-ui-common/components/controls/Dropdown';
import { PopupPlacement } from 'sonar-ui-common/components/ui/popups';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { getPathUrlAsString } from 'sonar-ui-common/helpers/urls';
import { getCodeUrl } from '../../../helpers/urls';
import { SourceViewerContext } from '../SourceViewerContext';

export interface LineOptionsPopupProps {
  firstLineNumber: number;
  line: T.SourceLine;
}

export function LineOptionsPopup({ firstLineNumber, line }: LineOptionsPopupProps) {
  return (
    <SourceViewerContext.Consumer>
      {({ branchLike, file }) => {
        const codeLocation = getCodeUrl(file.project, branchLike, file.key, line.line);
        const codeUrl = getPathUrlAsString(codeLocation, false);
        const isAtTop = line.line - 4 < firstLineNumber;
        return (
          <DropdownOverlay
            className="big-spacer-left"
            noPadding={true}
            placement={isAtTop ? PopupPlacement.BottomLeft : PopupPlacement.TopLeft}>
            <ul className="padded source-viewer-bubble-popup nowrap">
              <ActionsDropdownItem copyValue={codeUrl}>
                {translate('component_viewer.copy_permalink')}
              </ActionsDropdownItem>
            </ul>
          </DropdownOverlay>
        );
      }}
    </SourceViewerContext.Consumer>
  );
}

export default React.memo(LineOptionsPopup);
