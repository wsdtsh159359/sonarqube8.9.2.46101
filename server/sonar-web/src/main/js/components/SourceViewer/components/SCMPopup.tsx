/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import DateFormatter from 'sonar-ui-common/components/intl/DateFormatter';
import { translate } from 'sonar-ui-common/helpers/l10n';

export interface SCMPopupProps {
  line: T.SourceLine;
}

export function SCMPopup({ line }: SCMPopupProps) {
  const hasAuthor = line.scmAuthor !== undefined && line.scmAuthor !== '';
  const hasDate = line.scmDate !== undefined;
  return (
    <div className="source-viewer-bubble-popup abs-width-400">
      {hasAuthor && (
        <div>
          <h4>{translate('author')}</h4>
          {line.scmAuthor}
        </div>
      )}
      {hasDate && (
        <div className={classNames({ 'spacer-top': hasAuthor })}>
          <h4>{translate('source_viewer.tooltip.scm.commited_on')}</h4>
          <DateFormatter date={line.scmDate!} />
        </div>
      )}
      {line.scmRevision && (
        <div className={classNames({ 'spacer-top': hasAuthor || hasDate })}>
          <h4>{translate('source_viewer.tooltip.scm.revision')}</h4>
          {line.scmRevision}
        </div>
      )}
    </div>
  );
}

export default React.memo(SCMPopup);
