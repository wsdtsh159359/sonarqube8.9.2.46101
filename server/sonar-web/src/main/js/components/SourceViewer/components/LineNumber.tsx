/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Toggler from 'sonar-ui-common/components/controls/Toggler';
import { translateWithParameters } from 'sonar-ui-common/helpers/l10n';
import LineOptionsPopup from './LineOptionsPopup';

export interface LineNumberProps {
  displayOptions: boolean;
  firstLineNumber: number;
  line: T.SourceLine;
}

export function LineNumber({ displayOptions, firstLineNumber, line }: LineNumberProps) {
  const [isOpen, setOpen] = React.useState<boolean>(false);
  const { line: lineNumber } = line;
  const hasLineNumber = !!lineNumber;

  return hasLineNumber ? (
    <td className="source-meta source-line-number" data-line-number={lineNumber}>
      {displayOptions ? (
        <Toggler
          closeOnClickOutside={true}
          onRequestClose={() => setOpen(false)}
          open={isOpen}
          overlay={<LineOptionsPopup firstLineNumber={firstLineNumber} line={line} />}>
          <span
            aria-expanded={isOpen}
            aria-haspopup={true}
            aria-label={translateWithParameters('source_viewer.line_X', lineNumber)}
            onClick={() => setOpen(true)}
            role="button"
            tabIndex={0}>
            {lineNumber}
          </span>
        </Toggler>
      ) : (
        lineNumber
      )}
    </td>
  ) : (
    <td className="source-meta source-line-number" />
  );
}

export default React.memo(LineNumber);
