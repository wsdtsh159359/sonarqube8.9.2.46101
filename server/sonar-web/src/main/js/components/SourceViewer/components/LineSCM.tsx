/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Dropdown from 'sonar-ui-common/components/controls/Dropdown';
import { PopupPlacement } from 'sonar-ui-common/components/ui/popups';
import { translate, translateWithParameters } from 'sonar-ui-common/helpers/l10n';
import SCMPopup from './SCMPopup';

export interface LineSCMProps {
  line: T.SourceLine;
  previousLine: T.SourceLine | undefined;
}

export function LineSCM({ line, previousLine }: LineSCMProps) {
  const hasPopup = !!line.line;
  const cell = (
    <div className="source-line-scm-inner">
      {isSCMChanged(line, previousLine) ? line.scmAuthor || 'â€¦' : ' '}
    </div>
  );

  if (hasPopup) {
    let ariaLabel = translate('source_viewer.click_for_scm_info');
    if (line.scmAuthor) {
      ariaLabel = `${translateWithParameters(
        'source_viewer.author_X',
        line.scmAuthor
      )}, ${ariaLabel}`;
    }

    return (
      <td className="source-meta source-line-scm" data-line-number={line.line}>
        <Dropdown overlay={<SCMPopup line={line} />} overlayPlacement={PopupPlacement.RightTop}>
          <div aria-label={ariaLabel} role="button">
            {cell}
          </div>
        </Dropdown>
      </td>
    );
  } else {
    return (
      <td className="source-meta source-line-scm" data-line-number={line.line}>
        {cell}
      </td>
    );
  }
}

function isSCMChanged(s: T.SourceLine, p: T.SourceLine | undefined) {
  let changed = true;
  if (p != null && s.scmRevision != null && p.scmRevision != null) {
    changed = s.scmRevision !== p.scmRevision || s.scmDate !== p.scmDate;
  }
  return changed;
}

export default React.memo(LineSCM);
