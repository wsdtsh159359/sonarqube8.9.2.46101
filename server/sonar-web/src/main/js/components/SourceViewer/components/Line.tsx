/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import { times } from 'lodash';
import * as React from 'react';
import { BranchLike } from '../../../types/branch-like';
import './Line.css';
import LineCode from './LineCode';
import LineCoverage from './LineCoverage';
import LineDuplicationBlock from './LineDuplicationBlock';
import LineIssuesIndicator from './LineIssuesIndicator';
import LineNumber from './LineNumber';
import LineSCM from './LineSCM';

interface Props {
  branchLike: BranchLike | undefined;
  displayAllIssues?: boolean;
  displayCoverage: boolean;
  displayDuplications: boolean;
  displayIssueLocationsCount?: boolean;
  displayIssueLocationsLink?: boolean;
  displayIssues: boolean;
  displayLineNumberOptions?: boolean;
  displayLocationMarkers?: boolean;
  displaySCM?: boolean;
  duplications: number[];
  duplicationsCount: number;
  firstLineNumber: number;
  highlighted: boolean;
  highlightedLocationMessage: { index: number; text: string | undefined } | undefined;
  highlightedSymbols: string[] | undefined;
  issueLocations: T.LinearIssueLocation[];
  issuePopup: { issue: string; name: string } | undefined;
  issues: T.Issue[];
  last: boolean;
  line: T.SourceLine;
  loadDuplications: (line: T.SourceLine) => void;
  onIssueChange: (issue: T.Issue) => void;
  onIssuePopupToggle: (issueKey: string, popupName: string, open?: boolean) => void;
  onIssuesClose: (line: T.SourceLine) => void;
  onIssueSelect: (issueKey: string) => void;
  onIssuesOpen: (line: T.SourceLine) => void;
  onIssueUnselect: () => void;
  onLocationSelect: ((x: number) => void) | undefined;
  onSymbolClick: (symbols: string[]) => void;
  openIssues: boolean;
  previousLine: T.SourceLine | undefined;
  renderDuplicationPopup: (index: number, line: number) => React.ReactNode;
  scroll?: (element: HTMLElement) => void;
  scrollToUncoveredLine?: boolean;
  secondaryIssueLocations: T.LinearIssueLocation[];
  selectedIssue: string | undefined;
  verticalBuffer?: number;
}

const LINE_HEIGHT = 18;

export default class Line extends React.PureComponent<Props> {
  handleIssuesIndicatorClick = () => {
    if (this.props.openIssues) {
      this.props.onIssuesClose(this.props.line);
      this.props.onIssueUnselect();
    } else {
      this.props.onIssuesOpen(this.props.line);

      const { issues } = this.props;
      if (issues.length > 0) {
        this.props.onIssueSelect(issues[0].key);
      }
    }
  };

  render() {
    const {
      branchLike,
      displayAllIssues,
      displayCoverage,
      displayDuplications,
      displayIssueLocationsCount,
      displayIssueLocationsLink,
      displayLineNumberOptions,
      displayLocationMarkers,
      highlightedLocationMessage,
      displayIssues,
      displaySCM = true,
      duplications,
      duplicationsCount,
      firstLineNumber,
      highlighted,
      highlightedSymbols,
      issueLocations,
      issuePopup,
      issues,
      last,
      line,
      openIssues,
      previousLine,
      scrollToUncoveredLine,
      secondaryIssueLocations,
      selectedIssue,
      verticalBuffer
    } = this.props;

    const className = classNames('source-line', {
      'source-line-highlighted': highlighted,
      'source-line-filtered': line.isNew,
      'source-line-filtered-dark':
        displayCoverage &&
        (line.coverageStatus === 'uncovered' || line.coverageStatus === 'partially-covered'),
      'source-line-last': last === true
    });

    const bottomPadding = verticalBuffer ? verticalBuffer * LINE_HEIGHT : undefined;

    // default is true
    const displayOptions = displayLineNumberOptions !== false;

    return (
      <tr className={className} data-line-number={line.line}>
        <LineNumber displayOptions={displayOptions} firstLineNumber={firstLineNumber} line={line} />

        {displaySCM && <LineSCM line={line} previousLine={previousLine} />}
        {displayIssues && !displayAllIssues ? (
          <LineIssuesIndicator
            issues={issues}
            issuesOpen={openIssues}
            line={line}
            onClick={this.handleIssuesIndicatorClick}
          />
        ) : (
          <td className="source-meta source-line-issues" />
        )}

        {displayDuplications && (
          <LineDuplicationBlock
            blocksLoaded={duplicationsCount > 0}
            duplicated={Boolean(line.duplicated)}
            index={0}
            key={0}
            line={this.props.line}
            onClick={this.props.loadDuplications}
            renderDuplicationPopup={this.props.renderDuplicationPopup}
          />
        )}

        {duplicationsCount > 1 &&
          times(duplicationsCount - 1, index => (
            <LineDuplicationBlock
              blocksLoaded={true}
              duplicated={duplications.includes(index + 1)}
              index={index + 1}
              key={index + 1}
              line={this.props.line}
              renderDuplicationPopup={this.props.renderDuplicationPopup}
            />
          ))}

        {displayCoverage && (
          <LineCoverage
            line={line}
            scroll={this.props.scroll}
            scrollToUncoveredLine={scrollToUncoveredLine}
          />
        )}

        <LineCode
          branchLike={branchLike}
          displayIssueLocationsCount={displayIssueLocationsCount}
          displayIssueLocationsLink={displayIssueLocationsLink}
          displayLocationMarkers={displayLocationMarkers}
          highlightedLocationMessage={highlightedLocationMessage}
          highlightedSymbols={highlightedSymbols}
          issueLocations={issueLocations}
          issuePopup={issuePopup}
          issues={issues}
          line={line}
          onIssueChange={this.props.onIssueChange}
          onIssuePopupToggle={this.props.onIssuePopupToggle}
          onIssueSelect={this.props.onIssueSelect}
          onLocationSelect={this.props.onLocationSelect}
          onSymbolClick={this.props.onSymbolClick}
          padding={bottomPadding}
          scroll={this.props.scroll}
          secondaryIssueLocations={secondaryIssueLocations}
          selectedIssue={selectedIssue}
          showIssues={openIssues || displayAllIssues}
        />
      </tr>
    );
  }
}
