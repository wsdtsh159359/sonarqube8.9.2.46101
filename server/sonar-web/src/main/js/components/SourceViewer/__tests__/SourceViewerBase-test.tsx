/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import { getComponentData, getComponentForSourceViewer, getSources } from '../../../api/components';
import { mockMainBranch } from '../../../helpers/mocks/branch-like';
import { mockIssue, mockSourceLine, mockSourceViewerFile } from '../../../helpers/testMocks';
import defaultLoadIssues from '../helpers/loadIssues';
import SourceViewerBase from '../SourceViewerBase';

jest.mock('../helpers/loadIssues', () => ({
  default: jest.fn().mockRejectedValue({})
}));

jest.mock('../../../api/components', () => ({
  getComponentForSourceViewer: jest.fn().mockRejectedValue(''),
  getComponentData: jest.fn().mockRejectedValue(''),
  getSources: jest.fn().mockRejectedValue('')
}));

beforeEach(() => {
  jest.resetAllMocks();
});

it('should render nothing from the start', () => {
  expect(shallowRender().type()).toBeNull();
});

it('should render correctly', async () => {
  (defaultLoadIssues as jest.Mock).mockResolvedValueOnce([mockIssue()]);
  (getComponentForSourceViewer as jest.Mock).mockResolvedValueOnce(mockSourceViewerFile());
  (getComponentData as jest.Mock).mockResolvedValueOnce({
    component: { leakPeriodDate: '2018-06-20T17:12:19+0200' }
  });
  (getSources as jest.Mock).mockResolvedValueOnce([]);

  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  expect(wrapper).toMatchSnapshot();
});

it('should use load props if provided', () => {
  const loadComponent = jest.fn().mockResolvedValue({});
  const loadIssues = jest.fn().mockResolvedValue([]);
  const loadSources = jest.fn().mockResolvedValue([]);
  const wrapper = shallowRender({
    loadComponent,
    loadIssues,
    loadSources
  });

  expect(wrapper.instance().loadComponent).toBe(loadComponent);
});

it('should reload', async () => {
  (defaultLoadIssues as jest.Mock)
    .mockResolvedValueOnce([mockIssue()])
    .mockResolvedValueOnce([mockIssue()]);
  (getComponentForSourceViewer as jest.Mock).mockResolvedValueOnce(mockSourceViewerFile());
  (getComponentData as jest.Mock).mockResolvedValueOnce({
    component: { leakPeriodDate: '2018-06-20T17:12:19+0200' }
  });
  (getSources as jest.Mock).mockResolvedValueOnce([mockSourceLine()]);

  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().reloadIssues();

  expect(defaultLoadIssues).toBeCalledTimes(2);

  await waitAndUpdate(wrapper);

  expect(wrapper.state().issues).toHaveLength(1);
});

it('should load sources before', async () => {
  (defaultLoadIssues as jest.Mock)
    .mockResolvedValueOnce([mockIssue(false, { key: 'issue1' })])
    .mockResolvedValueOnce([mockIssue(false, { key: 'issue2' })]);
  (getComponentForSourceViewer as jest.Mock).mockResolvedValueOnce(mockSourceViewerFile());
  (getComponentData as jest.Mock).mockResolvedValueOnce({
    component: { leakPeriodDate: '2018-06-20T17:12:19+0200' }
  });
  (getSources as jest.Mock)
    .mockResolvedValueOnce([mockSourceLine()])
    .mockResolvedValueOnce([mockSourceLine()]);

  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().loadSourcesBefore();
  expect(wrapper.state().loadingSourcesBefore).toBe(true);

  expect(defaultLoadIssues).toBeCalledTimes(2);
  expect(getSources).toBeCalledTimes(2);

  await waitAndUpdate(wrapper);
  expect(wrapper.state().loadingSourcesBefore).toBe(false);
  expect(wrapper.state().issues).toHaveLength(2);
});

it('should load sources after', async () => {
  (defaultLoadIssues as jest.Mock)
    .mockResolvedValueOnce([mockIssue(false, { key: 'issue1' })])
    .mockResolvedValueOnce([mockIssue(false, { key: 'issue2' })]);
  (getComponentForSourceViewer as jest.Mock).mockResolvedValueOnce(mockSourceViewerFile());
  (getComponentData as jest.Mock).mockResolvedValueOnce({
    component: { leakPeriodDate: '2018-06-20T17:12:19+0200' }
  });
  (getSources as jest.Mock)
    .mockResolvedValueOnce([mockSourceLine()])
    .mockResolvedValueOnce([mockSourceLine()]);

  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().loadSourcesAfter();
  expect(wrapper.state().loadingSourcesAfter).toBe(true);

  expect(defaultLoadIssues).toBeCalledTimes(2);
  expect(getSources).toBeCalledTimes(2);

  await waitAndUpdate(wrapper);

  expect(wrapper.state().loadingSourcesAfter).toBe(false);
  expect(wrapper.state().issues).toHaveLength(2);
});

it('should handle no sources when checking ranges', () => {
  const wrapper = shallowRender();

  wrapper.setState({ sources: undefined });
  expect(wrapper.instance().isLineOutsideOfRange(12)).toBe(true);
});

function shallowRender(overrides: Partial<SourceViewerBase['props']> = {}) {
  return shallow<SourceViewerBase>(
    <SourceViewerBase branchLike={mockMainBranch()} component="my-component" {...overrides} />
  );
}
