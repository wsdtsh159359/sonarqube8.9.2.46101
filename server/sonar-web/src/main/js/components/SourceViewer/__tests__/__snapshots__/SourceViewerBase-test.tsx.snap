// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "branchLike": Object {
        "analysisDate": "2018-01-01",
        "excludedFromPurge": true,
        "isMain": true,
        "name": "master",
      },
      "file": Object {
        "key": "foo",
        "leakPeriodDate": "2018-06-20T17:12:19+0200",
        "measures": Object {
          "coverage": "85.2",
          "duplicationDensity": "1.0",
          "issues": "12",
          "lines": "56",
        },
        "path": "foo/bar.ts",
        "project": "my-project",
        "projectName": "MyProject",
        "q": "FIL",
        "uuid": "foo-bar",
      },
    }
  }
>
  <div
    className="source-viewer"
  >
    <ContextConsumer>
      <Component />
    </ContextConsumer>
    <SourceViewerCode
      branchLike={
        Object {
          "analysisDate": "2018-01-01",
          "excludedFromPurge": true,
          "isMain": true,
          "name": "master",
        }
      }
      componentKey="my-component"
      displayAllIssues={false}
      displayIssueLocationsCount={true}
      displayIssueLocationsLink={true}
      displayLocationMarkers={true}
      duplicationsByLine={Object {}}
      hasSourcesAfter={false}
      hasSourcesBefore={false}
      highlightedSymbols={Array []}
      issueLocationsByLine={
        Object {
          "25": Array [
            Object {
              "from": 0,
              "line": 25,
              "to": 999999,
            },
          ],
          "26": Array [
            Object {
              "from": 0,
              "line": 26,
              "to": 15,
            },
          ],
        }
      }
      issues={
        Array [
          Object {
            "actions": Array [],
            "component": "main.js",
            "componentLongName": "main.js",
            "componentQualifier": "FIL",
            "componentUuid": "foo1234",
            "creationDate": "2017-03-01T09:36:01+0100",
            "flows": Array [],
            "fromHotspot": false,
            "key": "AVsae-CQS-9G3txfbFN2",
            "line": 25,
            "message": "Reduce the number of conditional operators (4) used in the expression",
            "project": "myproject",
            "projectKey": "foo",
            "projectName": "Foo",
            "rule": "javascript:S1067",
            "ruleName": "foo",
            "secondaryLocations": Array [],
            "severity": "MAJOR",
            "status": "OPEN",
            "textRange": Object {
              "endLine": 26,
              "endOffset": 15,
              "startLine": 25,
              "startOffset": 0,
            },
            "transitions": Array [],
            "type": "BUG",
          },
        ]
      }
      issuesByLine={
        Object {
          "26": Array [
            Object {
              "actions": Array [],
              "component": "main.js",
              "componentLongName": "main.js",
              "componentQualifier": "FIL",
              "componentUuid": "foo1234",
              "creationDate": "2017-03-01T09:36:01+0100",
              "flows": Array [],
              "fromHotspot": false,
              "key": "AVsae-CQS-9G3txfbFN2",
              "line": 25,
              "message": "Reduce the number of conditional operators (4) used in the expression",
              "project": "myproject",
              "projectKey": "foo",
              "projectName": "Foo",
              "rule": "javascript:S1067",
              "ruleName": "foo",
              "secondaryLocations": Array [],
              "severity": "MAJOR",
              "status": "OPEN",
              "textRange": Object {
                "endLine": 26,
                "endOffset": 15,
                "startLine": 25,
                "startOffset": 0,
              },
              "transitions": Array [],
              "type": "BUG",
            },
          ],
        }
      }
      loadDuplications={[Function]}
      loadSourcesAfter={[Function]}
      loadSourcesBefore={[Function]}
      loadingSourcesAfter={false}
      loadingSourcesBefore={false}
      onIssueChange={[Function]}
      onIssuePopupToggle={[Function]}
      onIssueSelect={[Function]}
      onIssueUnselect={[Function]}
      onIssuesClose={[Function]}
      onIssuesOpen={[Function]}
      onSymbolClick={[Function]}
      openIssuesByLine={Object {}}
      renderDuplicationPopup={[Function]}
      sources={Array []}
      symbolsByLine={Object {}}
    />
  </div>
</ContextProvider>
`;
